synapse/storage/databases/main/appservice.py

```diff
@@ -285,40 +288,10 @@ async def complete_appservice_txn(
         """
         txn_id = int(txn_id)
 
-        def _complete_appservice_txn(txn):
-            # Debugging query: Make sure the txn being completed is EXACTLY +1 from
-            # what was there before. If it isn't, we've got problems (e.g. the AS
-            # has probably missed some events), so whine loudly but still continue,
-            # since it shouldn't fail completion of the transaction.
-            last_txn_id = self._get_last_txn(txn, service.id)
-            if (last_txn_id + 1) != txn_id:
-                logger.error(
-                    "appservice: Completing a transaction which has an ID > 1 from "
-                    "the last ID sent to this AS. We've either dropped events or "
-                    "sent it to the AS out of order. FIX ME. last_txn=%s "
-                    "completing_txn=%s service_id=%s",
-                    last_txn_id,
-                    txn_id,
-                    service.id,
-                )
-
-            # Set current txn_id for AS to 'txn_id'
-            self.db_pool.simple_upsert_txn(
-                txn,
-                "application_services_state",
-                {"as_id": service.id},
-                {"last_txn": txn_id},
-            )
```

Comment:
as a general rule, we'd generally prefer to make this sort of change over two releases of Synapse. In the first release, we'd stop reading from the `last_txn` column; we would then follow up in a second release by removing the code that *writes* to `last_txn`.  If we combine the changes into one release, it means that anybody who has a problem with the release and has to roll back will have problems (in this case, I think all the AS txns since upgrade will be replayed).

So for bonus points, it would be great to split this into two changes.

Reply:
V.sensible! https://github.com/matrix-org/synapse/pull/12209/commits/722f6ce13e86402c47ec58b3b24964f93851828b